File: ConvexHull.py
import numpy as np
import matplotlib.pyplot as plt

def main(alist):
    return_array= []                               # Forming final return array, empty to start with
    min_pointx = find_minx( alist )                # Finding point with min x coordinate
    max_pointx = find_maxx( alist )                # Finding point with max x coordinate
    return_array.append( min_pointx )            # Since these points will always be a part of the convex hull, add them to the return array
    return_array.append( max_pointx )   
    #plot(alist)
    
    #plt.plot(min_pointx, max_pointx)
    #plt.show()
    
    Upper_side = []        # Empty list of all the points 'above'/ on one side of the line drawn by min_pointx and maxpointx
    Lower_side = []        # Empty list of all the points 'below'/ on other side of the line drawn by min_pointx and maxpointx
    for Point in alist:
        Signed_area = tri_area( min_pointx, max_pointx, Point ) # Calculating signed area of triangle formed by some point, min_pointx and maxpointx
        #print ("Signed Area is " + str(Signed_area))
        if Signed_area > 0:
            Upper_side.append( Point )                       # Appending depending on sign of triangle
        if Signed_area < 0:
            Lower_side.append( Point )
    
    #print ("Upper Points are" + str(Upper_side))
    #print ("Lower Points are" + str(Lower_side))
            
    hull1 = hull( min_pointx, max_pointx, Upper_side, True )# Passing in min_pointx, maxpointx, array of points and a boolean for the +ve triangle area side
    #print (hull1)                                            # ....returns list of points
    hull2 = hull( min_pointx, max_pointx, Lower_side, False )
    #print (hull2)
    #print (return_array)
    return return_array + hull1 + hull2
        
    
    #print (Upper_side)
    #print (Lower_side)

def hull( PointA, PointB, ArrPoints, PositiveArea ):
    if ( len(ArrPoints) == 1 or len(ArrPoints) == 0 ):                         # Base Case
        return ArrPoints
    
    max_area = abs( tri_area( PointA, PointB, ArrPoints[0] ) )   # Finding out which point in ArrPonts is furthest from the line drawn by PointA and PointB 
    max_areaP = ArrPoints[ 0 ]
    for Point in ArrPoints:
        Area = abs( tri_area( PointA, PointB, Point ) )
        #print( "Abs Area is" + str(Area))
        if Area > max_area:
            max_area = Area
            max_areaP = Point
    
    #print ("Furthest Point is" + str(max_areaP))
    Recur_arrayA = []     # The triangle PointA, PointB, max_areaP divides arrPoints in 3 regions. Need to call hull recursively on the two
                          # regions that face outwards, i.e. points not contained in the triangle.
    Recur_arrayB = []
    for Point in ArrPoints:
        Signed_area = tri_area( PointA, max_areaP, Point )
        
         # if ArrPoints were the points with positive area to start with (PositiveArea == True), then the points in the
         # 2 regions of importance will have positive triangle area w.r.t PointA and max_areaP 
        
        if PositiveArea:                                                 
            if Signed_area > 0:
                Recur_arrayA.append(Point)
        
          # if ArrPoints were the points with positive area to start with (PositiveArea == True), then the points in the 
          # 2 regions of importance will have positive triangle area w.r.t PointA and max_areaP    
                
        else:                            
            if Signed_area < 0:
                Recur_arrayA.append(Point)
                
        Signed_area = tri_area( max_areaP, PointB, Point )  # Repeat with PointB
        #print ("Signed Area for PointB and maxAreaP is" + str(Signed_area))
        if PositiveArea:
            if Signed_area > 0:
                Recur_arrayB.append(Point)
        else:
            if Signed_area < 0:
                Recur_arrayB.append(Point)
    
    #print ("RecurArrA is" + str(Recur_arrayA))
    #print ("RecurArrB is" + str(Recur_arrayB))

    
    return [max_areaP] + hull(PointA, max_areaP, Recur_arrayA, PositiveArea) + hull(PointB, max_areaP, Recur_arrayB, PositiveArea)
        
    
# Helper Functions   

def find_minx(alist):
    min_pointx = alist[0]
    for x in alist:
        if x[0] < min_pointx[0]:
            min_pointx = x
    return min_pointx

def find_maxx(alist):
    max_pointx = alist[0]
    for x in alist:
        if x[0] > max_pointx[0]:
            max_pointx = x
    return max_pointx

def plot(alist):
    x_coords = []
    y_coords = []
    
    for point in alist:
        x_coords.append(point[0])
        y_coords.append(point[1])
    
    plt.plot(x_coords, y_coords, 'ro')
    plt.show()
    
def tri_area( PointA, PointB, PointC ):                             #calculates area of triangle spawned by PointA, PointB, PointC
    vectorAB = [PointB[0] - PointA[0], PointB[1] - PointA[1] ]      #Use vector calculus to caluclate signed area
    vectorAC = [PointC[0] - PointA[0], PointC[1] - PointA[1] ]
    CrossProduct_mag = vectorAB[0] * vectorAC[1] - vectorAB[1] * vectorAC[0]  #Cross Product
    
    return CrossProduct_mag
    

File: convexhull.cc
#include "convexhull.h"
#include <vector>
#include <stdlib.h>
#include <iostream>
//struct Point{
//	x_value;
//	y_value;
//};

// helper functions below this line //


//Given an rvector object with Points, returns the Point with minimum x value
Point find_minx( std::vector<Point>& alist ){
	std::vector<Point>::iterator begin = alist.begin();
	std::vector<Point>::iterator end = alist.end();
	Point min_pointx = *begin;
	while ( begin != end ){
		if ((*begin).x_value < min_pointx.x_value){
			min_pointx = *begin;
		}
		begin++;
	}
	return min_pointx;
}


//Given an rvector object with Points, returns the Point with maximum x value
Point find_maxx( std::vector<Point>& alist ){
	std::vector<Point>::iterator begin = alist.begin();
	std::vector<Point>::iterator end  = alist.end();
	Point min_pointx = *begin;
	while ( begin != end ){
		if ((*begin).x_value > min_pointx.x_value){
			min_pointx = *begin;
		}
		begin++;
	}
	return min_pointx;
}


//Given 3 Points, the funtion returns the area of the triangle spawned by the 3 points using the cross product.
//AB X AC -->  Sign of the result is determined through the right hand rule and the z axis.
int tri_area(Point pointA, Point pointB, Point pointC){
	int vectorAB_x = pointB.x_value - pointA.x_value;
	int vectorAB_y = pointB.y_value - pointA.y_value;
	int vectorAC_x = pointC.x_value - pointA.x_value;
	int vectorAC_y = pointC.y_value - pointA.y_value;
	
	int cross_product = ((vectorAB_x * vectorAC_y) - (vectorAB_y * vectorAC_x));
	return cross_product;
}  
	

int convexhull( std::vector<Point>& alist ){
	//forming the basic return array that will just contain the Points with the minimum and maximum x coordinate
	std::vector<Point> return_array;

	//find the Point with the least x coordinate and the Point with the max x coordinate; put in return array
	Point min_pointx = find_minx( alist );
	Point max_pointx = find_maxx( alist );
  return_array.push_back(min_pointx);
	return_array.push_back(max_pointx);	
 
	for (auto i: return_array){
		std::cout << "x value is" << i.x_value << "and y value is" << i.y_value;
	}

	//make 2 vectors, upper side and lower side; they contain points that lie above and below the line	
	//drawn by min_pointx and max_pointx as determined by the triangle area helper function

	std::vector<Point> upper_side;
	std::vector<Point> lower_side;
	for (auto i : alist){
		int signed_area = tri_area( min_pointx, max_pointx, i );
		if (signed_area > 0){
			upper_side.push_back( i );
 		}	
		else{
			lower_side.push_back( i );
		}
	}
	for (auto i: upper_side){
		std::cout << "x value of upper side  is" << i.x_value << "and y value of upper side is" << i.y_value << "\n";
	}
	
	for (auto i: lower_side){
		std::cout << "x value of lower side is" << i.x_value << "and y value of lower side is" << i.y_value << "\n";
	}		

		
	

	//std::vector<Point> hull1 = hull( min_pointx, max_pointx, upper_side, True );
	//std::vector<Point> hull2 = hull( min_pointx, max_pointx, lower_side, False );
	
//	return_array.insert( return_array.end(), hull1.begin(), hull1.end() )
//	return_array.insert( return_array.end(), hull2.begin(), hull2.end() )
	return 0;
}

int main(){
	Point point1;
	point1.x_value = 1;
	point1.y_value = 2;
	
	Point point2;
	point2.x_value = 3;
	point2.y_value = 1;
	
	Point point3;
	point3.x_value = 2;
	point3.y_value = 5;

	Point point4;
	point4.x_value = 6;
	point4.y_value = 2;

	std::vector<Point> myvector;
	myvector.push_back(point1);
	myvector.push_back(point2);
	myvector.push_back(point3);
	myvector.push_back(point4);

	convexhull(myvector);
	return 0;
}












	
